Add ability to delete difficulties from the editor

Gramore opened the pull request:
Addresses issue: https://github.com/ppy/osu/issues/19361
Honestly this was so painless to implement I didn't have time to get any questions out about design I believe it's in a reasonable state after a bit of fiddling.

Design Decision: Added popup for delete
This differs from the hide in the song list, but since this is in the editor it might be more worthy of the popup. Should this maybe only prompt if it's had changes or something or be removed all together?

As far as I can see no tests in this area and I wouldn't know where to begin adding them 😅. Let me know if there's a better way to implement this in this codebase, I'm fresh as so just kind of went at it.

New Menu:
![image](https://user-images.githubusercontent.com/40532209/181211476-5c24e075-dad7-4fe3-bb81-512a303a4818.png)
(not sure i like all the spacing)

Dialogue:
![image](https://user-images.githubusercontent.com/40532209/181212115-bb63380e-4958-4be0-9267-ef82a1fa88ed.png)
(Delete is a press and hold)

Gramore committed:
Adds delete difficulty option to editor

Gramore committed:
Dialogue touchups

frenzibyte commented on osu.Game/Screens/Edit/Editor.cs:
```diff
@@ -849,6 +849,20 @@ private void exportBeatmap()
+            var current = playableBeatmap.BeatmapInfo;
+            if (current is null) return;
+
+            beatmapManager.Hide(current);
```
Not sure about showing to the user that the difficulty will be "deleted" when it appears to be just hidden.

Gramore replied:
Agreed that is doesn't read well but the summary of the method `Hide` is "Delete a beatmap difficulty."
The `Delete` method appears to remove the whole beatmap. 
Happy to be steered on this one :)

Gramore replied:
Doh I think I messed up my filter when playing with delete, I'll take another crack.

Gramore replied:
Actually sorry for the noise but currently the `Delete` operates on `BeatmapSetInfo` not `BeatmapInfo` (my bad for not reading)
Is an implementation required for  `BeatmapInfo`? I don't want to go fiddling with stuff that I shouldn't without some guidance :)

frenzibyte replied:
I would wait for a second opinion but I personally think difficulties should be purged/soft-deleted in editor rather than hidden (which could become an annoyance if a user creates random diffs and want to delete them, especially when we have flow for showing hidden beatmaps in song select, or if the users uses the "Restore all hidden difficulties" button in settings).

Gramore replied:
Potentially adding a 'Hide instead' option or something to the dialogue here if that's desired but you are making a lot of sense.

peppy replied:
Hiding is the terminology and methodology used because I don't want users to be able to delete beatmaps from a set. This will break things going forward (such as comparing a beatmap set with its online version – we don't want random files to be missing).

I'll have to have a think about how this should be handled in the editor, but probably a hard delete with ample confirmation is okay for now.

Gramore replied:
Would you like me to have a crack at adding a hard delete for beatmaps to the manager then or hold off while you have a think?
Any more confirmation than the current implementations dialogue? (2 stage / are you really sure?)

peppy replied:
Go for it. It should be pretty simple.

Please make sure to add test coverage. And make sure the editor loads another difficulty from the same beatmap if available.

Gramore committed:
Adds hard-delete for a beatmap diff

Gramore commented on osu.Game/Beatmaps/BeatmapManager.cs:
```diff
@@ -358,6 +358,31 @@ public void Delete(Expression<Func<BeatmapSetInfo, bool>>? filter = null, bool s
+                    if (!beatmapInfo.IsManaged)
+                        beatmapInfo = r.Find<BeatmapInfo>(beatmapInfo.ID);
+
+                    var setInfo = beatmapInfo.BeatmapSet!;
+                    DeleteFile(setInfo, beatmapInfo.File!);
+                    setInfo.Beatmaps.Remove(beatmapInfo);
+
+                    var liveSetInfo = r.Find<BeatmapSetInfo>(setInfo.ID);
+                    setInfo.CopyChangesToRealm(liveSetInfo);
```
Can someone confirm that this is reasonable?

Gramore commented on osu.Game/Screens/Edit/PromptForDifficultyDeleteDialog.cs:
```diff
@@ -0,0 +1,38 @@
+            {
+                new PopupDialogOkButton
+                {
+                    Text = @"Hide this difficulty instead (recommended)",
```
Happy with this phrasing?

peppy replied:
I'm not sure this makes sense. If you are a mapper, hiding a difficulty is something you'd basically never choose.

Gramore commented on osu.Game/Screens/Edit/Editor.cs:
```diff
@@ -849,6 +849,40 @@ private void exportBeatmap()
+                this.Exit();
+            var nextBeatmap = setInfo!.Beatmaps.First();
+
+            // Force a refresh of the beatmap (and beatmap set) so the `Change difficulty` list is also updated.
+            Beatmap.Value = beatmapManager.GetWorkingBeatmap(nextBeatmap, true);
+            SwitchToDifficulty(Beatmap.Value.BeatmapInfo);
```
Sanity check here as well.

peppy replied:
These comments are unnecessary. We review every line.

Gramore replied:
Apologies, spent too much time around rubber stamps 😅

Gramore commented:
Some guidance on how to go about adding tests for this would be excellent!

peppy commented:
> Some guidance on how to go about adding tests for this would be excellent!

Check out https://github.com/ppy/osu/blob/6220650ea3228930ebd27a46b9e4c8c3824abc73/osu.Game.Tests/Visual/Editing/TestSceneEditorSaving.cs#L23

Feodor0090 commented:
Is this PR abandoned? I wrote a test for this.

<details>

```
public class TestSceneDifficultyDelete : EditorTestScene
    {
        protected override Ruleset CreateEditorRuleset() => new OsuRuleset();
        protected override bool IsolateSavingFromDatabase => false;

        [Resolved]
        private OsuGameBase game { get; set; } = null!;

        [Resolved]
        private BeatmapManager beatmaps { get; set; } = null!;

        private BeatmapSetInfo importedBeatmapSet = null!;

        protected override WorkingBeatmap CreateWorkingBeatmap(IBeatmap beatmap, Storyboard storyboard = null!)
            => beatmaps.GetWorkingBeatmap(importedBeatmapSet.Beatmaps.First());

        public override void SetUpSteps()
        {
            AddStep("import test beatmap", () => importedBeatmapSet = BeatmapImportHelper.LoadOszIntoOsu(game, virtualTrack: true).GetResultSafely());
            base.SetUpSteps();
        }

        [Test]
        public void TestDifficultyDelete()
        {
            string lastDiff = null!;
            AddStep("remember selected difficulty", () => lastDiff = EditorBeatmap.BeatmapInfo.DifficultyName);

            AddStep("click File", () => this.ChildrenOfType<DrawableOsuMenuItem>().First().TriggerClick());
            AddStep("click Delete", () => this.ChildrenOfType<DrawableOsuMenuItem>().Single(deleteMenuItemPredicate).TriggerClick());
            AddStep("confirm", () => InputManager.Key(Key.Number2));

            AddAssert("difficulty is deleted", () =>
            {
                if (lastDiff == null!) throw new NullReferenceException();

                var newSet = beatmaps.GetWorkingBeatmap(importedBeatmapSet.Beatmaps.First(), true).BeatmapSetInfo;
                return newSet.Beatmaps.All(x => x.DifficultyName != lastDiff);
            });
        }

        private bool deleteMenuItemPredicate(DrawableOsuMenuItem item)
        {
            return item.ChildrenOfType<SpriteText>().Any(text => text.Text.ToString().StartsWith("Delete", StringComparison.Ordinal));
        }
    }
```
</details>

peppy committed:
Merge branch 'master' into feature/EditorDeleteDiff

peppy committed:
Add test scene

peppy committed:
Refactor/rename deletion method to read better

peppy committed:
Tidy up dialog and deletion flow

peppy commented:
> Is this PR abandoned?

No it has not been (by me at least), but it is still in a not great state.

> I wrote a test for this.

Thanks, I've added this. Although your tests only tests the passing case, and not the multiple failures:

- Might crash depending on which difficulty is deleted due to a flaw in the select-next-beatmap logic.
- Allows deleting the last difficulty which leaves a `BetmapSet` orphaned with no difficulties.

Gramore commented:
Ah yes I got quite stuck getting the tests going and very much forgot it had this on my plate.
I think there was a bug with handling non-saved beatmaps that i hadn't ironed out yet, but aside from that happy if you want to pick it up from here but I'll take another pass tomorrow though with your tests as a base (thank you!)

peppy committed:
Disable delete difficulty menu item when only one difficulty is present

peppy committed:
Ensure next beatmap selected matches the menu ordering

peppy committed:
Add extra coverage to tests

Also fixes a potential failure due to test beatmap having two difficulties with same name.

peppy commented:
I've fixed the issues I had with this.

Requesting a second review.

frenzibyte commented on osu.Game/Screens/Edit/Editor.cs:
```diff
@@ -879,35 +878,61 @@ private void seek(UIEvent e, int direction)
 
-            fileMenuItems.Add(createDifficultyCreationMenu());
-            fileMenuItems.Add(createDifficultySwitchMenu());
+                Beatmap.Value = beatmapManager.GetWorkingBeatmap(nextToShow, true);
```
`SwitchToDifficulty` already updates the game-wide beatmap bindable, but I suppose this is done to invalidate the working beatmap as well?

I think it would make more sense to invalidate directly and do it inside `DeleteDifficultyImmediately` as well, similar to `CreateNewDifficulty`/`CopyExistingDifficulty` (invalidating beatmapset at the end of method).

peppy replied:
I've moved the invalidation, but this call still needs to exist to advance to the correct difficulty post-deletion.

frenzibyte commented on osu.Game/Beatmaps/BeatmapManager.cs:
```diff
@@ -363,6 +363,30 @@ public void Delete(Expression<Func<BeatmapSetInfo, bool>>? filter = null, bool s
+                Debug.Assert(beatmapInfo.BeatmapSet != null);
+                Debug.Assert(beatmapInfo.File != null);
+
+                var setInfo = beatmapInfo.BeatmapSet;
+
+                DeleteFile(setInfo, beatmapInfo.File);
+                setInfo.Beatmaps.Remove(beatmapInfo);
```
This is missing updating beatmapset hash and changing status to `LocallyModified`, but not 100% sure if this should also invoke `ProcessBeatmap` as well.

peppy replied:
I don't think calling `ProcessBeatmap` is required, but I've added the hash change, good call.

frenzibyte reviewed (changes requested):
Resorting to review as I'm not quite familiar with the database flow and Realm to resolve the concerns myself and merge this.

peppy committed:
Fix deleting a difficulty not updating the beatmap set hash

peppy committed:
Invalidate working beatmap cache when calling `DeleteDifficultyImmediately` rather than in editor code

frenzibyte commented:
Deleting a difficulty and exiting without saving any beatmap does not show the update button:

Uploading CleanShot 2022-09-06 at 17.56.22.mp4…

This is due to the carousel beatmap code solely relying on `BeatmapSetInfo.AllBeatmapsUpToDate`, which remains true when deleting a difficulty since it doesn't keep track of it.

Simplest way I can think of to solve this would be to enforce the update button when `Status == LocallyAvailable`, @peppy thoughts?

peppy commented:
That might be worth opening an issue for. I would feel uncomfortable trying to fix in this PR.

frenzibyte committed:
Merge branch 'master' into feature/EditorDeleteDiff
